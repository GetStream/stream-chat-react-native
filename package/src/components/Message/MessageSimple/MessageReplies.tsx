import React from 'react';
import { ColorValue, Platform, StyleSheet, Text, TouchableOpacity, View } from 'react-native';

import {
  MessageContextValue,
  useMessageContext,
} from '../../../contexts/messageContext/MessageContext';
import {
  MessagesContextValue,
  useMessagesContext,
} from '../../../contexts/messagesContext/MessagesContext';
import { useTheme } from '../../../contexts/themeContext/ThemeContext';
import {
  TranslationContextValue,
  useTranslationContext,
} from '../../../contexts/translationContext/TranslationContext';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    flexDirection: 'row',
    marginTop: 8,
  },
  curveContainer: {
    flexDirection: 'row',
  },
  leftMessageRepliesCurve: {
    borderBottomLeftRadius: 16,
    borderRightColor: 'transparent',
    ...Platform.select({
      android: {
        borderRightWidth: 0,
      },
    }),
  },
  messageRepliesCurve: {
    borderTopColor: 'transparent',
    borderTopWidth: 0,
    borderWidth: 1,
    height: 16,
    width: 16,
  },
  messageRepliesText: {
    fontSize: 12,
    fontWeight: '700',
    paddingBottom: 5,
    paddingLeft: 8,
  },
  rightMessageRepliesCurve: {
    borderBottomRightRadius: 16,
    borderLeftColor: 'transparent',
    ...Platform.select({
      android: {
        borderLeftWidth: 0,
      },
    }),
  },
});

export type MessageRepliesPropsWithContext<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = Pick<
  MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>,
  | 'alignment'
  | 'message'
  | 'onLongPress'
  | 'onPress'
  | 'onPressIn'
  | 'onOpenThread'
  | 'preventPress'
  | 'threadList'
> &
  Pick<MessagesContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'MessageRepliesAvatars'> &
  Pick<TranslationContextValue, 't'> & {
    noBorder?: boolean;
    repliesCurveColor?: ColorValue;
  };

const MessageRepliesWithContext = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  props: MessageRepliesPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    alignment,
    message,
    MessageRepliesAvatars,
    noBorder,
    onLongPress,
    onOpenThread,
    onPress,
    onPressIn,
    preventPress,
    repliesCurveColor,
    t,
    threadList,
  } = props;

  const {
    theme: {
      colors: { accent_blue },
      messageSimple: {
        replies: { container, leftCurve, messageRepliesText, rightCurve },
      },
    },
  } = useTheme();

  if (threadList || !message.reply_count) return null;

  return (
    <View style={styles.curveContainer}>
      {alignment === 'left' && (
        <>
          {!noBorder && (
            <View
              style={[
                { borderColor: repliesCurveColor },
                styles.messageRepliesCurve,
                styles.leftMessageRepliesCurve,
                leftCurve,
              ]}
            />
          )}
          <MessageRepliesAvatars alignment={alignment} message={message} />
        </>
      )}
      <TouchableOpacity
        disabled={preventPress}
        onLongPress={(event) => {
          if (onLongPress) {
            onLongPress({
              emitter: 'messageReplies',
              event,
            });
          }
        }}
        onPress={(event) => {
          if (onPress) {
            onPress({
              defaultHandler: onOpenThread,
              emitter: 'messageReplies',
              event,
            });
          }
        }}
        onPressIn={(event) => {
          if (onPressIn) {
            onPressIn({
              defaultHandler: onOpenThread,
              emitter: 'messageReplies',
              event,
            });
          }
        }}
        style={[styles.container, container]}
        testID='message-replies'
      >
        <Text style={[styles.messageRepliesText, { color: accent_blue }, messageRepliesText]}>
          {message.reply_count === 1
            ? t('1 Thread Reply')
            : t('{{ replyCount }} Thread Replies', {
                replyCount: message.reply_count,
              })}
        </Text>
      </TouchableOpacity>
      {alignment === 'right' && (
        <>
          <MessageRepliesAvatars alignment={alignment} message={message} />
          {!noBorder && (
            <View
              style={[
                { borderColor: repliesCurveColor },
                styles.messageRepliesCurve,
                styles.rightMessageRepliesCurve,
                rightCurve,
              ]}
            />
          )}
        </>
      )}
    </View>
  );
};

const areEqual = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  prevProps: MessageRepliesPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
  nextProps: MessageRepliesPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    message: prevMessage,
    noBorder: prevNoBorder,
    t: prevT,
    threadList: prevThreadList,
  } = prevProps;
  const {
    message: nextMessage,
    noBorder: nextNoBorder,
    t: nextT,
    threadList: nextThreadList,
  } = nextProps;

  const threadListEqual = prevThreadList === nextThreadList;
  if (!threadListEqual) return false;

  const messageReplyCountEqual = prevMessage.reply_count === nextMessage.reply_count;
  if (!messageReplyCountEqual) return false;

  const noBorderEqual = prevNoBorder === nextNoBorder;
  if (!noBorderEqual) return false;

  const tEqual = prevT === nextT;
  if (!tEqual) return false;

  return true;
};

const MemoizedMessageReplies = React.memo(
  MessageRepliesWithContext,
  areEqual,
) as typeof MessageRepliesWithContext;

export type MessageRepliesProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = Partial<MessageRepliesPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>>;

export const MessageReplies = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  props: MessageRepliesProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    alignment,
    message,
    onLongPress,
    onOpenThread,
    onPress,
    onPressIn,
    preventPress,
    threadList,
  } = useMessageContext<At, Ch, Co, Ev, Me, Re, Us>();
  const { MessageRepliesAvatars } = useMessagesContext<At, Ch, Co, Ev, Me, Re, Us>();
  const { t } = useTranslationContext();

  return (
    <MemoizedMessageReplies
      {...{
        alignment,
        message,
        MessageRepliesAvatars,
        onLongPress,
        onOpenThread,
        onPress,
        onPressIn,
        preventPress,
        t,
        threadList,
      }}
      {...props}
    />
  );
};

MessageReplies.displayName = 'MessageReplies{messageSimple{replies}}';
