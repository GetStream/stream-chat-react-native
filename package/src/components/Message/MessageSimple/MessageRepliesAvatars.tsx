import React from 'react';
import { StyleSheet, View } from 'react-native';

import { ChatContextValue, useChatContext } from '../../../contexts/chatContext/ChatContext';
import type { MessageContextValue } from '../../../contexts/messageContext/MessageContext';
import { useTheme } from '../../../contexts/themeContext/ThemeContext';

import type { DefaultStreamChatGenerics } from '../../../types/types';
import { Avatar } from '../../Avatar/Avatar';

const styles = StyleSheet.create({
  avatarContainer: {
    alignItems: 'center',
    flexDirection: 'row',
    paddingTop: 2,
  },
  topAvatar: {
    paddingTop: 2,
    position: 'absolute',
  },
});

export type MessageRepliesAvatarsProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Pick<MessageContextValue<StreamChatGenerics>, 'alignment' | 'message'>;

export const MessageRepliesAvatarsWithContext = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  props: MessageRepliesAvatarsProps<StreamChatGenerics> & Pick<ChatContextValue, 'ImageComponent'>,
) => {
  const { alignment, ImageComponent, message } = props;

  const {
    theme: {
      colors: { white_snow },
      messageSimple: {
        replies: {
          avatar,
          avatarContainerMultiple,
          avatarContainerSingle,
          avatarSize,
          leftAvatarsContainer,
          rightAvatarsContainer,
        },
      },
    },
  } = useTheme();

  const avatars = message.thread_participants?.slice(-2) || [];
  const hasMoreThanOneReply = avatars.length > 1;

  return (
    <View
      style={[
        styles.avatarContainer,
        alignment === 'right' ? { marginLeft: 8, ...rightAvatarsContainer } : leftAvatarsContainer,
      ]}
    >
      {avatars.map((user, i) => (
        <View
          key={user.id}
          style={
            i === 1
              ? { ...styles.topAvatar, ...avatarContainerSingle }
              : {
                  paddingLeft: hasMoreThanOneReply ? 8 : 0,
                  ...avatarContainerMultiple,
                }
          }
        >
          <Avatar
            containerStyle={[
              i === 1 && {
                borderColor: white_snow,
                borderWidth: 1,
              },
              avatar,
            ]}
            image={user.image}
            ImageComponent={ImageComponent}
            name={user.name}
            size={avatarSize ? avatarSize : i === 1 ? 18 : 16}
          />
        </View>
      ))}
    </View>
  );
};

export const MessageRepliesAvatars = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  props: MessageRepliesAvatarsProps<StreamChatGenerics>,
) => {
  const { ImageComponent } = useChatContext();

  return <MessageRepliesAvatarsWithContext {...props} ImageComponent={ImageComponent} />;
};
