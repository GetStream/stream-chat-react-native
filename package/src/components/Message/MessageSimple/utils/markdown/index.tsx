import React, { PropsWithChildren, useMemo } from 'react';
import { View } from 'react-native';

import SimpleMarkdown, {
  OutputRules,
  ParserRules,
  ReactOutputRule,
} from '@khanacademy/simple-markdown';
import { isArray, isEqual, merge } from 'lodash';

import { getLocalRules } from './rules';
import styles from './styles';

import { MarkdownStyle } from '../../../../../contexts';

type DefaultRules = typeof SimpleMarkdown.defaultRules;

export type MarkdownRules = Partial<DefaultRules>;

export type MarkdownProps = {
  onLink: (url: string) => Promise<void>;
  rules: MarkdownRules;
  styles: MarkdownStyle;
};

export type MarkdownOptions = Partial<Pick<MarkdownProps, 'onLink'>>;

const Markdown = (props: PropsWithChildren<MarkdownProps>) => {
  const { onLink, rules: rulesProp, styles: stylesProp, children } = props;

  const mergedStyles = useMemo(() => merge({}, styles, stylesProp), [stylesProp]);
  const localRules = useMemo(
    () =>
      merge(
        {},
        SimpleMarkdown.defaultRules,
        { ...SimpleMarkdown.defaultRules },
        getLocalRules(mergedStyles, { onLink }),
        rulesProp,
      ) as unknown as ParserRules,
    [mergedStyles, onLink, rulesProp],
  );

  const parser = useMemo(() => SimpleMarkdown.parserFor(localRules), [localRules]);
  const renderer = useMemo(
    () => SimpleMarkdown.outputFor(localRules as unknown as OutputRules<ReactOutputRule>, 'react'),
    [localRules],
  );

  const childText = useMemo(() => (isArray(children) ? children.join('') : children), [children]);

  const toRender = useMemo(() => {
    const blockSource = `${childText ?? ''}\n\n`;
    return parser(blockSource, { inline: false });
  }, [childText, parser]);

  const tree = useMemo(() => renderer(toRender), [renderer, toRender]);

  return <View style={[styles.view, stylesProp?.view]}>{tree}</View>;
};

const areEqual = (prevProps: PropsWithChildren, nextProps: PropsWithChildren) =>
  isEqual(prevProps.children, nextProps.children);

export default React.memo(Markdown, areEqual);
