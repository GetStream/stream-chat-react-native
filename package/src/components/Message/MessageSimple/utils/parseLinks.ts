/**
 * ðŸ“£ Note: Do not use named capture groups, as
 *          it is not yet available in Hermes.
 *
 * These helpers indend to make it easier to skim read
 * the patterns below, but note that some are optional,
 * specified in the patterns these are added to.
 * */
const emailUserName = '[\\w+\\.~$_-]+';
const schema = `(\\w{2,15}:\\/\\/)`;
const domain = `((?:\\w+\\.\\w+)+(?:[^:\\/\\s]+))`;
const port = `(:[0-9]{1,5})`;
const path = `((?:\\/)?[^?#\\s]+)`;
const queryString = `(\\?[^#\\s]+)`;
const fragment = `(#[\\w_-]+)`;

/**
 * Match any email address, with and without `mailto:`
 */
const emailPattern = `(mailto:)?((?:${emailUserName})@(?:${domain}))`;

/**
 * Match any string starting with something that seems like it's a schema.
 * */
const schemePrefixedLinkPattern = `${schema}(\\S+)`;

/**
 * Match as much as possible of a fqdn, at least with a domain name formed as
 * something.tld
 */
const fqdnLinkPattern = `${schema}?${domain}${port}?${path}?${queryString}?${fragment}?`;

interface Link {
  encoded: string;
  raw: string;
  scheme: string;
}

/**
 * This is done separately because of the version of javascript run
 * for expo
 * */
const removeMarkdownLinksFromText = (input: string) => input.replace(/\[[\w\s]+\]\(.*\)/g, '');

export const parseLinksFromText = (input: string): Link[] => {
  let matches;

  const inputWithoutMarkdownLinks = removeMarkdownLinksFromText(input);

  const results: Link[] = [];

  const emailRegExp = new RegExp(emailPattern, 'gi');
  while ((matches = emailRegExp.exec(input)) !== null) {
    const [raw, scheme = 'mailto:', displayValue] = matches;
    results.push({ encoded: encodeURI(displayValue), raw, scheme });
  }

  /**
   * The two link patterns are checked with an "or" (`|`)
   * to avoid overlapping matches being duplicated in the output.
   * */
  const linkRegex = new RegExp(`${fqdnLinkPattern}|${schemePrefixedLinkPattern}`, 'gi');
  while ((matches = linkRegex.exec(inputWithoutMarkdownLinks)) !== null) {
    const [
      raw,
      fqdnScheme = '',
      fqdnDomainName = '',
      fqdnPort = '',
      fqdnPath = '',
      fqdnQueryStraing = '',
      fqdnFragment = '',
      schemePrefixedScheme = '',
      schemePrefixedPath = '',
    ] = matches;

    if (schemePrefixedScheme !== '') {
      results.push({
        encoded: encodeURI(schemePrefixedPath),
        raw,
        scheme: schemePrefixedScheme,
      });
      continue;
    }

    results.push({
      encoded: encodeURI(
        [fqdnDomainName, fqdnPort, fqdnPath, fqdnQueryStraing, fqdnFragment].join(''),
      ),
      raw,
      scheme: fqdnScheme,
    });
  }

  return results;
};
