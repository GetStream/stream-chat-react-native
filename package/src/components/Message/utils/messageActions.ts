import type { MessageContextValue } from '../../../contexts/messageContext/MessageContext';
import type { OwnCapabilitiesContextValue } from '../../../contexts/ownCapabilitiesContext/OwnCapabilitiesContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import type { MessageActionType } from '../../MessageOverlay/MessageActionListItem';

export type MessageActionsParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = {
  blockUser: MessageActionType;
  copyMessage: MessageActionType;
  deleteMessage: MessageActionType;
  dismissOverlay: () => void;
  editMessage: MessageActionType;
  error: boolean | Error;
  flagMessage: MessageActionType;
  isThreadMessage: boolean;
  messageReactions: boolean;
  muteUser: MessageActionType;
  ownCapabilities: OwnCapabilitiesContextValue;
  pinMessage: MessageActionType;
  quotedReply: MessageActionType;
  retry: MessageActionType;
  threadReply: MessageActionType;
  unpinMessage: MessageActionType;
} & Pick<MessageContextValue<At, Ch, Co, Ev, Me, Re, Us>, 'message' | 'isMyMessage'>;

export type MessageActionsProp<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
> = (param: MessageActionsParams<At, Ch, Co, Ev, Me, Re, Us>) => MessageActionType[];

export const messageActions = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>({
  blockUser,
  copyMessage,
  deleteMessage,
  editMessage,
  error,
  flagMessage,
  isMyMessage,
  isThreadMessage,
  message,
  messageReactions,
  ownCapabilities,
  pinMessage,
  quotedReply,
  retry,
  threadReply,
  unpinMessage,
}: MessageActionsParams<At, Ch, Co, Ev, Me, Re, Us>) => {
  if (messageReactions) {
    return undefined;
  }

  const actions: Array<MessageActionType | null> = [];

  if (error && isMyMessage) {
    actions.push(retry);
  }

  if (ownCapabilities.quoteMessage && !isThreadMessage && !error) {
    actions.push(quotedReply);
  }

  if (ownCapabilities.sendReply && !isThreadMessage && !error) {
    actions.push(threadReply);
  }

  if (
    (isMyMessage && ownCapabilities.updateOwnMessage) ||
    (!isMyMessage && ownCapabilities.updateAnyMessage)
  ) {
    actions.push(editMessage);
  }

  if (message.text && !error) {
    actions.push(copyMessage);
  }

  if (!isMyMessage && ownCapabilities.flagMessage) {
    actions.push(flagMessage);
  }

  if (ownCapabilities.pinMessage && !message.pinned) {
    actions.push(pinMessage);
  }

  if (ownCapabilities.pinMessage && message.pinned) {
    actions.push(unpinMessage);
  }

  if (!isMyMessage && ownCapabilities.banChannelMembers) {
    actions.push(blockUser);
  }

  if (
    (isMyMessage && ownCapabilities.deleteOwnMessage) ||
    (!isMyMessage && ownCapabilities.deleteAnyMessage)
  ) {
    actions.push(deleteMessage);
  }

  return actions;
};
