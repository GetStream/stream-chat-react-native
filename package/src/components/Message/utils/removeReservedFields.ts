import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';
import type { MessageType } from '../../MessageList/hooks/useMessageList';

export const removeReservedFields = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType,
>(
  message: MessageType<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const retryMessage = { ...message };
  const reserved = [
    'cid',
    'config',
    'created_at',
    'created_by',
    'id',
    'last_message_at',
    'member_count',
    'type',
    'updated_at',
  ];
  reserved.forEach((key) => {
    delete retryMessage[key];
  });

  return retryMessage;
};
