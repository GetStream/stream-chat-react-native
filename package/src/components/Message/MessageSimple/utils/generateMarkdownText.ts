import truncate from 'lodash/truncate';

import { parseLinksFromText } from './parseLinks';

import type { DefaultStreamChatGenerics } from '../../../../types/types';
import type { MessageType } from '../../../MessageList/hooks/useMessageList';

export const generateMarkdownText = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  message: MessageType<StreamChatGenerics>,
) => {
  const { text } = message;

  if (!text) return null;

  // Trim the extra spaces from the text.
  let resultText = text.trim();

  // List of all the links present in the text.
  const linkInfos = parseLinksFromText(resultText);

  for (const linkInfo of linkInfos) {
    const displayLink = truncate(linkInfo.raw, {
      length: 200,
      omission: '...',
    });
    // Convert raw links/emails in the text to respective markdown syntax.
    // Eg: Hi getstream.io -> Hi [getstream.io](getstream.io).
    const normalRegEx = new RegExp(linkInfo.raw, 'g');
    const markdown = `[${displayLink}](${linkInfo.encodedUrl})`;
    resultText = text.replace(normalRegEx, markdown);

    // After previous step, in some cases, the mentioned user after `@` might have a link/email so we convert it back to normal raw text.
    // Eg: Hi, @[test.user@gmail.com](mailto:test.user@gmail.com) to @test.user@gmail.com.
    const mentionsRegex = new RegExp(`@\\[${displayLink}\\]\\(${linkInfo.encodedUrl}\\)`, 'g');
    resultText = resultText.replace(mentionsRegex, `@${displayLink}`);
  }

  resultText = resultText.replace(/[<&"'>]/g, '\\$&');

  return resultText;
};
