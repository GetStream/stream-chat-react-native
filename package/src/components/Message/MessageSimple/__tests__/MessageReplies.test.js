import React from 'react';

import { cleanup, fireEvent, render, waitFor } from '@testing-library/react-native';


import { ThemeProvider } from '../../../../contexts/themeContext/ThemeContext';
import { defaultTheme } from '../../../../contexts/themeContext/utils/theme';
import { TranslationProvider } from '../../../../contexts/translationContext/TranslationContext';
import { generateMessage } from '../../../../mock-builders/generator/message';
import { generateStaticUser, generateUser } from '../../../../mock-builders/generator/user';
import { MessageReplies } from '../MessageReplies';
import { MessageRepliesAvatars } from '../MessageRepliesAvatars';

afterEach(cleanup);

describe('MessageReplies', () => {
  it('should render message replies', async () => {
    const onPressMock = jest.fn(() => null);
    const t = jest.fn((key) => key);
    const staticUser = generateStaticUser(0);
    const message = generateMessage({
      reply_count: 2,
      user: staticUser,
    });
    const { getByTestId, getByText, queryAllByTestId, rerender, toJSON } = render(
      <TranslationProvider value={{ t }}>
        <ThemeProvider style={defaultTheme}>
          <MessageReplies
            alignment='right'
            groupStyles={['bottom']}
            message={message}
            MessageRepliesAvatars={MessageRepliesAvatars}
            openThread={onPressMock}
          />
        </ThemeProvider>
      </TranslationProvider>,
    );

    await waitFor(() => {
      expect(getByTestId('message-replies')).toBeTruthy();
      expect(getByTestId('message-replies-right')).toBeTruthy();
      expect(queryAllByTestId('message-replies-left')).toHaveLength(0);
      expect(t).toHaveBeenCalledWith('{{ replyCount }} Thread Replies', {
        replyCount: message.reply_count,
      });
      expect(getByText('{{ replyCount }} Thread Replies')).toBeTruthy();
    });

    const message2 = generateMessage({
      reply_count: 1,
      user: staticUser,
    });

    rerender(
      <TranslationProvider value={{ t }}>
        <ThemeProvider style={defaultTheme}>
          <MessageReplies
            alignment='left'
            groupStyles={['bottom']}
            message={message2}
            MessageRepliesAvatars={MessageRepliesAvatars}
            onPress={onPressMock}
          />
        </ThemeProvider>
      </TranslationProvider>,
    );

    fireEvent.press(getByTestId('message-replies'));

    await waitFor(() => {
      expect(onPressMock).toHaveBeenCalled();
      expect(getByTestId('message-replies')).toBeTruthy();
      expect(getByTestId('message-replies-left')).toBeTruthy();
      expect(queryAllByTestId('message-replies-right')).toHaveLength(0);
      expect(t).toHaveBeenCalledWith('1 Thread Reply');
      expect(getByText('1 Thread Reply')).toBeTruthy();
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should not render message replies', async () => {
    const t = jest.fn((key) => key);
    const user = generateUser();
    const message = generateMessage({
      user,
    });
    const { queryAllByTestId, rerender } = render(
      <TranslationProvider value={{ t }}>
        <ThemeProvider style={defaultTheme}>
          <MessageReplies
            alignment='right'
            groupStyles={['bottom']}
            message={message}
            onPress={() => null}
          />
        </ThemeProvider>
      </TranslationProvider>,
    );

    await waitFor(() => {
      expect(queryAllByTestId('message-replies')).toHaveLength(0);
      expect(queryAllByTestId('message-replies-left')).toHaveLength(0);
      expect(queryAllByTestId('message-replies-right')).toHaveLength(0);
    });

    const message2 = generateMessage({
      reply_count: 1,
      user,
    });

    rerender(
      <TranslationProvider value={{ t }}>
        <ThemeProvider style={defaultTheme}>
          <MessageReplies
            alignment='right'
            groupStyles={['bottom']}
            message={message2}
            onPress={() => null}
            threadList
          />
        </ThemeProvider>
      </TranslationProvider>,
    );

    await waitFor(() => {
      expect(queryAllByTestId('message-replies')).toHaveLength(0);
      expect(queryAllByTestId('message-replies-left')).toHaveLength(0);
      expect(queryAllByTestId('message-replies-right')).toHaveLength(0);
    });
  });
});
