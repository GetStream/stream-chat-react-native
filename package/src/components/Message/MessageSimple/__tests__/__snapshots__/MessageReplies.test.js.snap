// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageReplies should render message replies 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
    testID="message-replies-left"
  >
    <View
      style={
        Array [
          Object {
            "borderColor": undefined,
          },
          Object {
            "borderTopWidth": 0,
            "borderWidth": 1,
            "height": 16,
            "width": 16,
          },
          Object {
            "borderBottomLeftRadius": 16,
            "borderRightWidth": 0,
          },
          Object {},
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingTop": 2,
          },
          Object {},
        ]
      }
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginTop": 8,
        "opacity": 1,
      }
    }
    testID="message-replies"
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 12,
            "fontWeight": "700",
            "paddingBottom": 5,
            "paddingHorizontal": 8,
          },
          Object {
            "color": "#005FFF",
          },
          Object {},
        ]
      }
    >
      1 Thread Reply
    </Text>
  </View>
</View>
`;
